PyTorch NNAPI integration prototype

dreiss opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47407 Make bundled inputs work with quantized zero inputs
* **#46780 PyTorch NNAPI integration prototype**
* #46781 Add a command-line flag for overriding pthreadpool size

Summary:
This is in prototype status, but pretty functional.  There are two major
parts.

- Model converter.  This is a pure Python component that consumes a
  model in TorchScript format, converts the operations into NNAPI
  semantics, and serializes the model in a custom format.  It then wraps
  the result in a new TorchScript model that can invoke NNAPI under the
  hood.
- Runtime.  This is a TorchBind object that deserializes the model and
  sends the result to NNAPI.  This is fairly simple since the serialized
  format is basically just a list of NNAPI calls to make, so most of the
  code is spent on bounds checking.

A few notes on the design.
- Currently, all tensor sizes need to be fixed, and those fixed sizes
  are burned directly into the serialized model.  This will probably
  need to change.  NNAPI supports variable-sized tensors, but the
  important hardware backends do not.  However, we're seeing use cases
  crop up where the input size is not known until around the time that
  the model is loaded (for example, it might depend on the camera aspect
  ratio).  I think the proper fix here is to remove the code in the
  converter that eagerly calculates the sizes of the intermediate
  tensors and replace it with a code generator that will generate some
  TorchScript code that will perform those calculations at model load
  time.  This way, we will be able to support models that have
  variable-sized inputs while still only showing fixed-sized operands to
  NNAPI.
- The important hardware backends want operands to be in NHWC order, but
  PyTorch natively represents all tensors and NCHW.  The strategy for
  this is to keep NCHW during most of the conversion process, but track
  and additional value per operand representing the "dimension order".
  The dimension order gets propagated through convolutions and pointwise
  ops.  When we're ready to serialize the model, we reorder the
  dimensions for "channels last" operands to NHWC.

Test Plan:
Some local testing with FB prod models.  I'll need to add some examples
and automated tests.

Differential Revision: [D24574040](https://our.internmc.facebook.com/intern/diff/D24574040)

dhruvbird commented on aten/src/ATen/nnapi/nnapi_bind.cpp:
```diff
@@ -0,0 +1,197 @@
+void load_platform_library() {
+  static int run_once = [](){
+    nnapi_wrapper_load(&nnapi, &check_nnapi);
+    CAFFE_ENFORCE(nnapi);
+    CAFFE_ENFORCE(nnapi->Model_free);
+    CAFFE_ENFORCE(nnapi->Compilation_free);
+    CAFFE_ENFORCE(nnapi->Execution_free);
```
Q. Why do you enforce that only 3 functions exist and not the others?

dreiss replied:
The other functions are all called through the `check_nnapi` interface, which throws if the functions are null.  However, these need to be called from destructors, where throwing is not safe, so I call the raw NNAPI functions directly, so I want to verify early that they are valid.

dhruvbird commented on aten/src/ATen/nnapi/NeuralNetworks.h:
```diff
@@ -0,0 +1,84 @@
+    int32_t zeroPoint;
+} ANeuralNetworksOperandType;
+
+#endif  // MINIMAL_NEURAL_NETWORKS_H
```
Okay so this file is mainly from NNAPIs source tree - i.e. enum and struct definitions.

dhruvbird commented on torch/backends/_nnapi/prepare.py:
```diff
@@ -0,0 +1,167 @@
+            if fmt == 0:
+                fixed_args.append(args[idx].contiguous())
+            elif fmt == 1:
+                fixed_args.append(args[idx].permute(0,2,3,1).contiguous())
```
Okay this is where the NHCW stuff you mentioned earlier happens.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+    TENSOR_QUANT16_ASYMM = 12
+
+
+class NNAPI_OperationCode(object):
```
Okay, these were picked up from https://developer.android.com/ndk/reference/group/neural-networks#operationcode

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+        return struct.pack("i" * len(ints), *ints)
+
+
+    ADDER_MAP = {
```
Okay so these are the aten operators supported for the NNAPI execution. I'm guessing this list is dictated by the operation codes for the operators from https://developer.android.com/ndk/reference/group/neural-networks#operationcode

dreiss replied:
Yeah, this maps from the internal TorchScript operator name to the function that needs to handle that node.  In some cases (like processing constants), there's no NNAPI operation generated.

iseeyuan replied:
I'm assuming this map could grow if we are supporting more and more nnapi operations?

dreiss replied:
Yes.  Though I'd like to replace the explicit map with a decorator.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+class NNAPI_FuseCode(object):
+    FUSED_NONE = 0
+    FUSED_RELU = 1
+    FUSED_RELU1 = 2
+    FUSED_RELU6 = 3
```
Seems like none of the current calls reference RELU1 and RELU6.

dreiss replied:
Correct.  I just added them all because the list was short.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+        if config is None:
+            config = {}
+
+        self.solid_weights = config.get("solid_weights", False)
```
Q. What is `solid_weights`?

dreiss replied:
Oh, this is a vestigal feature.  I'll delete it.  Currently, we get each weight from a separate tensor.  "solid weights" was to let all of the weights be bundled as a single blob to make it possible to deploy without using PyTorch or Caffe2 as a wrapper.

dhruvbird replied:
Thanks! Now I understand the intent, but not the details, so I shan't mull over them for now.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+        self.modules = {}
+        self.constants = {}
+        self.jitval_operand_map = {}
+        self.cached_immediates = {}
```
Okay - `cached_immediates` is a set of all literal constant values that show up in this model, de-duped.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+    return abs(lhs - rhs) <= tolerance * min(lhs, rhs)
+
+
+def tensor_size(op_type, dims):
```
Okay so this is the tensor size in bytes.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+    assert len(shape2) > 0
+    s1 = list(shape1)
+    s2 = list(shape2)
+    if len(s1) > len(s2):
+        #s2 = [1] * (len(s1) - len(s2)) + s2
+        raise Exception("Non-equal-rank broadcast is too dangerous because XXX.")
+    if len(s2) > len(s1):
+        #s3 = [1] * (len(s2) - len(s1)) + s1
+        raise Exception("Non-equal-rank broadcast is too dangerous because XXX.")
```
@dreiss Maybe this is not relevant here, but https://pytorch.org/docs/stable/notes/broadcasting.html suggests that broadcasting can account for non-equal shapes of tensors. Is this something that would be a limitation of this specific API?

dreiss replied:
I need to clean this code up a bit.  In theory, PyTorch and NNAPI have the same broadcasting semantics (back to front, extend front with 1s).  However, when using "nnapi_nhwc", it gets a bit wonky, because PyTorch uses an implicit NHWC representation based on strides, so the tensors still have NCHW semantics, and broadcasting order is still W,H,C,N.  NNAPI (and TF and C2) use explicit NHWC, so the broadcast order is C,W,H,N.  The trickiest bits come into play when we try to broadcast a known NHWC tensors with a constant (where we don't necessarily known the user's intention).  I haven't worked through all the cases yet, so I decided to not support non-equal-rank broadcast for now.

dhruvbird replied:
Thanks! I understand now that there is additional complexity when you're dealing with different orders of dimensions and that broadcasting only works when the dimensions are ordered the same, and at a high level I gather that you can't make that assertion here, so you've left this open. If you somehow knew the order of dimensions then you could have supported the 1 padding behaviour. Got it!

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+        model = []
+
+        version = 1
+        header = struct.pack(
+                "iiiiii",
+                version,
+                len(self.operands),
+                len(self.values),
+                len(self.operations),
+                len(self.inputs),
+                len(self.outputs),
+                )
+        model.append(header)
+
+        serialized_values, serialized_value_data = self.serialize_values()
+
+        model.extend(struct.pack("iifi", t, len(d), s, z) for (t,d,_m,s,z) in self.operands)
+        model.extend(serialized_values)
+        model.extend(struct.pack("iii", *x) for x in self.operations)
+        model.extend(self.serialize_ints(fix_shape(dims, mf)) for (_, dims, mf, _, _) in self.operands)
+        model.extend(serialized_value_data)
+        model.append(self.serialize_ints(self.operation_args))
+        model.append(self.serialize_ints(self.inputs))
+        model.append(self.serialize_ints(self.outputs))
```
Okay so the serialized model basically is a file which has the model header specifying version, num(operands), num(values), etc... where operands, values, etc... are fixed size structures and the de-serializer basically reads them off as such because it knows how many of each to expect.

dhruvbird commented on torch/backends/_nnapi/prepare.py:
```diff
@@ -0,0 +1,167 @@
+        delist_outptus = False
+
+    ser_model, used_weights, inp_mem_fmts, out_mem_fmts = serialize_model(model, nnapi_inputs)
+    ser_model_tensor = torch.tensor(list(ser_model), dtype=torch.uint8)
```
I don't understand this bit. The serialized model is a character stream (string) right? What is the purposes of converting it into a Tensor?

dreiss replied:
I trust PyTorch to do a better job serializing int8 tensors than strings.

dhruvbird commented on torch/backends/_nnapi/prepare.py:
```diff
@@ -0,0 +1,167 @@
+
+
+def convert_model_to_nnapi(model, inputs):
+    model = torch.jit.freeze(model)
```
(for my reference) Okay - it seems like freeze is a way to inline things. https://pytorch.org/docs/stable/generated/torch.jit.freeze.html

dreiss replied:
Yes, this eliminates a *lot* of TorchScript cruft so we don't have to understand it during conversion.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+        self.outputs = []
+
+        self.modules = {}
+        self.constants = {}
```
Maybe I'm missing something, but I don't see any place that constants is being set (or updated). I only see queries.

dhruvbird replied:
nvm I see it now as `self.constants[key]`.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+        self.operations.append((opcode, len(inputs), len(outputs)))
+        self.operation_args.extend(inputs + outputs)
+
+    def add_constant_value(self, jitval, ctype, value):
+        assert jitval not in self.constants
+        self.constants[jitval] = (ctype, value)
```
Q. Where can I find out more about the permissible values in `jitval` and what they mean?

dreiss replied:
This is a "Value" in the JIT IR.  https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.h#L148  It represents a local variable or the result of a TorchScript expression.

dhruvbird replied:
I don't see a conversion from `struct Value` to `std::string` or `const char*` in the file `csrc/jit/ir/ir.h`. How does one use `jitval` as a string in this context?

iseeyuan commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+        value = getattr(obj, name)
+        output = node.outputsAt(0)
+        ctype = output.type()
+        self.add_constant_value(output, ctype, value)
```
Good idea to freeze the model! So in nnapi models, there's no concept of "parameters"? The weights and bias are all "constants"? Does that mean it does not have training capability?

dreiss replied:
NNAPI is only for inference.

iseeyuan replied:
> > Would this model only be used in platforms with nnapi available (Android)?
> 
> Yes. (Though it is possible in theory to build the NNAPI CPU implementation for Linux and run the model directly, which is how I've been testing.)

Does that mean we may keep multiple versions of a model?
For Android, we deploy an nnapi model.
For iOS, I'm wondering if there is corresponding API layer (CoreML?)

dhruvbird replied:
@iseeyuan for Metal, I believe it's integrated as a separate backend instead. I'm also curious to know why NNAPI isn't implemented as a separate backend. One thought I had is that PyTorch represents image Tensors as NCHW whereas NNAP (based on @dreiss diff summary) requires them to be in NHWC format, so implementing it as a separate backend may have required the conversion on every operator entry/exit? Not sure.

iseeyuan commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+                self.add_getattr(node),
+            "prim::Constant": lambda self, node:
+                self.add_constant_node(node),
+            "prim::ListConstruct": lambda self, node:
```
Could nnapi has corresponding prim::DictConstruct, TupleConstruct, etc? Do you have any plan to handle those nodes?

dreiss replied:
NNAPI has no concepts of lists and dicts.  It only operates on tensors and scalars: https://android.googlesource.com/platform/frameworks/ml/+/refs/tags/android-11.0.0_r8/nn/runtime/include/NeuralNetworks.h#68

iseeyuan replied:
Like how you handled prim::ListConstruct, we may also write code to handle DictConstruct and TupleConstruct from torchscript side?

dreiss replied:
Oh, yes, I suppose so.

iseeyuan commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+
+
+def serialize_model(module, inputs, config=None):
+    return _NnapiSerializer(config).serialize_model(module, inputs)
```
Is inputs used inside _NnapiSerializer?

dreiss replied:
Yes.  We use it to determine the input shape.

iseeyuan commented on aten/src/ATen/nnapi/NeuralNetworks.h:
```diff
@@ -0,0 +1,84 @@
+    ANEURALNETWORKS_PREFER_LOW_POWER = 0,
+    ANEURALNETWORKS_PREFER_FAST_SINGLE_ANSWER = 1,
+    ANEURALNETWORKS_PREFER_SUSTAINED_SPEED = 2,
+} PreferenceCode;
```
It might be useful to guide the build with PreferenceCode. Should it be specified per model? What could I find the usage?

dreiss replied:
Yeah, I would like to expose this.  I haven't figured out a good API for it yet.

iseeyuan commented:
Awesome work on model serialization and nnapi binding! In addition to the embedded comments, I have two general questions: 
1. Looks like the serializer would convert the torchscript graph to nnapi model (a torch.nn.Module) and script it. Would this model only be used in platforms with nnapi available (Android)? 
2. If we deploy this nnapi model to production, do you estimate any blocker to convert it to bytecode?
3. Is there a test covering the serialization -> loading -> running, and verify the results are the same with the original torchscript model?

dreiss commented on aten/src/ATen/nnapi/codegen.py:
```diff
@@ -0,0 +1,155 @@
+            }
+            """)
+        .replace("__DEFINE_CHECK_FUNCTIONS__", "\n".join(define_checks))
+        .replace("__LOAD_FUNCTIONS__", "\n".join(load_functions))
```
The wrapper obviously has some size in the binary, but dlsym happens at runtime.

iseeyuan replied:
Got it. So we assume that libneuralnetworks.so is available if on-device.

dreiss commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+    TENSOR_QUANT16_ASYMM = 12
+
+
+class NNAPI_OperationCode(object):
```
Each NNAPI op defines its own semantics.  For example, Conv2D has two forms, one with 10 args and one with 13: https://android.googlesource.com/platform/frameworks/ml/+/refs/tags/android-11.0.0_r8/nn/runtime/include/NeuralNetworks.h#401 .  But if you look closer, some of those args are "Available since API level 29", so there are also a 7-arg version and a *different* 10-arg version.  I believe the two different 10-arg versions are distinguished by the types of the arguments.

iseeyuan replied:
I see. We may need to find the bast schema match of Conv2D manually, when converting a conv2d node to nnapi::Conv2D

dreiss commented:
> Would this model only be used in platforms with nnapi available (Android)?

Yes.  (Though it is possible in theory to build the NNAPI CPU implementation for Linux and run the model directly, which is how I've been testing.)

> If we deploy this nnapi model to production, do you estimate any blocker to convert it to bytecode?
I have tested with the lite interpreter.  There are some minor blockers, but I think I have appropriate fixes ready.

> Is there a test covering the serialization -> loading -> running, and verify the results are the same with the original torchscript model?

Unfortunately, it is currently only possible to run these models on Android.  I will push harder on getting a host build of NNAPI so we can automate this.

dhruvbird commented:
Regarding unit tests: @dreiss Is the NNAPI similator API written by you or available in open source? If so could it be checked in and tests run against that library. One option could be to have a differently code-generated copy of `nnapi_wrapper.cpp` which just calls into the emulation later methods instead of via the function pointers.


I have reviewed the code related to the general serialization and deserialization of the model in the NNAPI format and it seems fine. I checked if the case statements have some copy-paste issue, and it doesn't seem that they do. I would love to see some test coverage as @iseeyuan  mentioned, but I guess there are practical limitations? Specifically, `nnapi_wrapper`, `nnapi_model_loader`, `codegen`, and parts of `serializer` seem fine.

I haven't reviewed the code which actually translates the call from PyTorch into the NNAPI backend, since it requires some understanding of how the translation takes place, and I'm not certain if I have that understanding yet. This includes most code in `serializer`, which does the actual translation from pt->nnapi.

dreiss commented:
> Is the NNAPI similator API written by you or available in open source?

No.  I have a heavily hacked-up version of the Android source tree that is able to build libneuralnetworks.so on Linux.  I've asked the Android team about getting official support for this, but it is quite difficult.  I will continue to ask them.

dhruvbird commented on torch/backends/_nnapi/prepare.py:
```diff
@@ -0,0 +1,167 @@
+        fixed_args = []
+        for idx in range(len(args)):
+            fmt = self.inp_mem_fmts[idx]
+            if fmt == 0:
+                fixed_args.append(args[idx].contiguous())
+            elif fmt == 1:
+                fixed_args.append(args[idx].permute(0,2,3,1).contiguous())
```
Would it be feasible to comment here saying that `fmt == 0` is NCHW (i.e. channels first) and `fmt == 1` is NHWC or maybe used a enum for this?

dreiss replied:
Enum might be tricky since this needs to be TorchScript-compatible, but I can certainly add a comment.

dhruvbird commented on torch/backends/_nnapi/prepare.py:
```diff
@@ -0,0 +1,167 @@
+        return outs[0]
+
+
+def convert_model_to_nnapi(model, inputs):
+    model = torch.jit.freeze(model)
```
It's interesting that converting a model to NNAPI requires providing inputs and running the model? Am I reading this correctly?

dhruvbird replied:
nvm I think this just calls the constructor of the frozen model, and doesn't actually run it.

dreiss replied:
Currently, we only run the model in order to get the number, shape, dtype, and qparams of the outputs.  We could also do this without running the model.

dhruvbird commented on torch/backends/_nnapi/prepare.py:
```diff
@@ -0,0 +1,167 @@
+            out_mem_fmts: List[int],
+            out_templates: List[torch.Tensor]):
+        super().__init__()
+        self.ser_model = ser_model
```
@dreiss Okay I think I now understand why you did this. Your serialized model (for NNAPI) is just a Tensor<int8> represented as a regular member of the original model so that it will get naturally serialized when the model is serialized to torchscript. Is that right?

dreiss replied:
Yes.

dhruvbird commented on aten/src/ATen/nnapi/nnapi_bind.cpp:
```diff
@@ -0,0 +1,197 @@
+      buffer_sizes.push_back(t.nbytes());
+    }
+
+    TORCH_CHECK(serialized_model_tensor.is_contiguous());
+    c10::ArrayRef<uint8_t> ser_model = {
+      serialized_model_tensor.data_ptr<uint8_t>(),
+      serialized_model_tensor.nbytes()
+    };
```
Okay this is how it's read in.

dhruvbird commented on aten/src/ATen/nnapi/nnapi_model_loader.cpp:
```diff
@@ -0,0 +1,264 @@
+  if (len % 4 == 0) {
+    return len;
+  }
+  return len + 4 - (phys % 4);
```
Unlikely that this will result in an overflow, but it could if your model has size 4G - 3 bytes or more.

dreiss replied:
True, but I think we're a long way away from shipping a 4G model on NNAPI.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+            return value
+        raise Exception(f"Can't handle size arg of type '{ctype!r}' for '{jitval!r}'")
+
+    def expand_sizes(self, size):
+        return [ s.item() for s in size ]
```
This method seems to be unused.

dreiss replied:
Probably left-over from the old Caffe2 implementation.

dhruvbird commented on torch/backends/_nnapi/serializer.py:
```diff
@@ -0,0 +1,1131 @@
+
+            # Pad with 0 bytes out to a multiple of 4 for alignment.
+            physical_length = ((source_length - 1) | 0x3) + 1
+            padded_data = data + (b"\0" * (physical_length - source_length))
```
@dreiss Q. Does this assume a certain endianness?

dreiss replied:
I don't think so.  I assume little-endian somewhere else, though.

iseeyuan reviewed (approved):
LGTM in general. Is there a guidance on how to convert and deploy an nnapi model with an example? That may help me to run the code and dive a little deeper. I think it would be good for this first version in prototype status.

dreiss commented on torch/backends/_nnapi/prepare.py:
```diff
@@ -0,0 +1,167 @@
+    def forward(self, args: List[torch.Tensor]) -> List[torch.Tensor]:
+        return self.nnapi_module(args)
+    
+    @torch.jit.export
+    def __getstate__(self):
+        return self.nnapi_module
+        
+    @torch.jit.export
+    def __setstate__(self, nnapi_module):
+        self.training = False
+        self.nnapi_module = nnapi_module
+        self.nnapi_module.init()
```
The purpose here is to make sure that init gets called when the model is loaded, without the user having to intervene.

dreiss committed:
Update on "PyTorch NNAPI integration prototype"


Summary:
This is in prototype status, but pretty functional.  There are two major
parts.

- Model converter.  This is a pure Python component that consumes a
  model in TorchScript format, converts the operations into NNAPI
  semantics, and serializes the model in a custom format.  It then wraps
  the result in a new TorchScript model that can invoke NNAPI under the
  hood.
- Runtime.  This is a TorchBind object that deserializes the model and
  sends the result to NNAPI.  This is fairly simple since the serialized
  format is basically just a list of NNAPI calls to make, so most of the
  code is spent on bounds checking.

A few notes on the design.
- Currently, all tensor sizes need to be fixed, and those fixed sizes
  are burned directly into the serialized model.  This will probably
  need to change.  NNAPI supports variable-sized tensors, but the
  important hardware backends do not.  However, we're seeing use cases
  crop up where the input size is not known until around the time that
  the model is loaded (for example, it might depend on the camera aspect
  ratio).  I think the proper fix here is to remove the code in the
  converter that eagerly calculates the sizes of the intermediate
  tensors and replace it with a code generator that will generate some
  TorchScript code that will perform those calculations at model load
  time.  This way, we will be able to support models that have
  variable-sized inputs while still only showing fixed-sized operands to
  NNAPI.
- The important hardware backends want operands to be in NHWC order, but
  PyTorch natively represents all tensors and NCHW.  The strategy for
  this is to keep NCHW during most of the conversion process, but track
  and additional value per operand representing the "dimension order".
  The dimension order gets propagated through convolutions and pointwise
  ops.  When we're ready to serialize the model, we reorder the
  dimensions for "channels last" operands to NHWC.

Test Plan:
Some local testing with FB prod models.  I'll need to add some examples
and automated tests.

Differential Revision: [D24574040](https://our.internmc.facebook.com/intern/diff/D24574040)

[ghstack-poisoned]

dreiss committed:
Update on "PyTorch NNAPI integration prototype"


Summary:
This is in prototype status, but pretty functional.  There are two major
parts.

- Model converter.  This is a pure Python component that consumes a
  model in TorchScript format, converts the operations into NNAPI
  semantics, and serializes the model in a custom format.  It then wraps
  the result in a new TorchScript model that can invoke NNAPI under the
  hood.
- Runtime.  This is a TorchBind object that deserializes the model and
  sends the result to NNAPI.  This is fairly simple since the serialized
  format is basically just a list of NNAPI calls to make, so most of the
  code is spent on bounds checking.

A few notes on the design.
- Currently, all tensor sizes need to be fixed, and those fixed sizes
  are burned directly into the serialized model.  This will probably
  need to change.  NNAPI supports variable-sized tensors, but the
  important hardware backends do not.  However, we're seeing use cases
  crop up where the input size is not known until around the time that
  the model is loaded (for example, it might depend on the camera aspect
  ratio).  I think the proper fix here is to remove the code in the
  converter that eagerly calculates the sizes of the intermediate
  tensors and replace it with a code generator that will generate some
  TorchScript code that will perform those calculations at model load
  time.  This way, we will be able to support models that have
  variable-sized inputs while still only showing fixed-sized operands to
  NNAPI.
- The important hardware backends want operands to be in NHWC order, but
  PyTorch natively represents all tensors and NCHW.  The strategy for
  this is to keep NCHW during most of the conversion process, but track
  and additional value per operand representing the "dimension order".
  The dimension order gets propagated through convolutions and pointwise
  ops.  When we're ready to serialize the model, we reorder the
  dimensions for "channels last" operands to NHWC.

Test Plan:
Some local testing with FB prod models.  I'll need to add some examples
and automated tests.

Differential Revision: [D24574040](https://our.internmc.facebook.com/intern/diff/D24574040)

[ghstack-poisoned]

dreiss committed:
Update on "PyTorch NNAPI integration prototype"


Summary:
This is in prototype status, but pretty functional.  There are two major
parts.

- Model converter.  This is a pure Python component that consumes a
  model in TorchScript format, converts the operations into NNAPI
  semantics, and serializes the model in a custom format.  It then wraps
  the result in a new TorchScript model that can invoke NNAPI under the
  hood.
- Runtime.  This is a TorchBind object that deserializes the model and
  sends the result to NNAPI.  This is fairly simple since the serialized
  format is basically just a list of NNAPI calls to make, so most of the
  code is spent on bounds checking.

A few notes on the design.
- Currently, all tensor sizes need to be fixed, and those fixed sizes
  are burned directly into the serialized model.  This will probably
  need to change.  NNAPI supports variable-sized tensors, but the
  important hardware backends do not.  However, we're seeing use cases
  crop up where the input size is not known until around the time that
  the model is loaded (for example, it might depend on the camera aspect
  ratio).  I think the proper fix here is to remove the code in the
  converter that eagerly calculates the sizes of the intermediate
  tensors and replace it with a code generator that will generate some
  TorchScript code that will perform those calculations at model load
  time.  This way, we will be able to support models that have
  variable-sized inputs while still only showing fixed-sized operands to
  NNAPI.
- The important hardware backends want operands to be in NHWC order, but
  PyTorch natively represents all tensors and NCHW.  The strategy for
  this is to keep NCHW during most of the conversion process, but track
  and additional value per operand representing the "dimension order".
  The dimension order gets propagated through convolutions and pointwise
  ops.  When we're ready to serialize the model, we reorder the
  dimensions for "channels last" operands to NHWC.

Test Plan:
Some local testing with FB prod models.  I'll need to add some examples
and automated tests.

Differential Revision: [D24574040](https://our.internmc.facebook.com/intern/diff/D24574040)

[ghstack-poisoned]
